synapse/rest/client/v2_alpha/sync.py

```diff
@@ -324,6 +334,52 @@ async def encode_invited(self, rooms, time_now, token_id, event_formatter):
 
         return invited
 
+    async def encode_knocked(
+        self,
+        rooms: List[KnockedSyncResult],
+        time_now: int,
+        token_id: int,
+        event_formatter: Callable[[Dict], Dict],
+    ) -> Dict[str, Dict[str, Any]]:
+        """
+        Encode the rooms we've knocked on in a sync result.
+
+        Args:
+            rooms: list of sync results for rooms this user is knocking on
+            time_now: current time - used as a baseline for age calculations
+            token_id: ID of the user's auth token - used for namespacing of transaction IDs
+            event_formatter: function to convert from federation format to client format
+
+        Returns:
+            The list of rooms the user has knocked on, in our response format.
+        """
+        knocked = {}
+        for room in rooms:
+            knock = await self._event_serializer.serialize_event(
+                room.knock, time_now, token_id=token_id, event_format=event_formatter,
+            )
+
+            # Extract the `unsigned` key from the knock event.
+            # This is where we (cheekily) store the knock state events
+            unsigned = knock.setdefault("unsigned", {})
+
+            # Extract the stripped room state from the unsigned dict
+            # This is for clients to get a little bit of information about
+            # the room they've knocked on, without revealing any sensitive information
+            knocked_state = list(unsigned.pop("knock_room_state", []))
+
+            # Append the actual knock membership event itself as well
+            # TODO: I *believe* this is just for the client's sake of track its membership
+            # state in each room, but I could be wrong. This certainly doesn't seem like it
+            # could have any negative effects besides resource usage
```

Comment:
If this is the case, then it should probably be mentioned as a requirement in the MSC...

Reply:
Looks like we should figure this out?

Reply:
So, this was [discussed in #synapse-dev](https://matrix.to/#/!XaqDhxuTIlvldquJaV:matrix.org/$SPSdsZbVtS4z2Og8su_RmqRUthUzstMny2b0WjWMddI?via=matrix.org&via=vector.modular.im&via=vinix.im) a little while ago, and the conclusion was that the invite is included to give some extra membership information about the inviter to the invitee. Such that if the room you're invited to has no name and/or no avatar, it would show the inviter's avatar and name instead.

Now looking back at this, this doesn't make a whole lot of sense. An invite is an update to the membership state of the *invitee*, not the *inviter*, so really the invitee is just getting their own membership information sent to them again.

However, I suppose you would have the `sender` field in the invite, which may let you query profile information and thus get the name/avatar of the inviter. I don't know if this is what clients do in practice however.

Regardless though, with knocking the membership information and the `sender` field are all of the knocker, meaning including this knock event wouldn't give you any more information about the room than you already knew.

I've sent out a question to the Element teams to see if they'd need this for any other reason.

Reply:
And the response from Element Web and Mobile is that it's not necessary. It was questioned whether this would still allow getting knock state on an initialSync, and the answer is yes, your `/sync` stream just may look like:

```json
{
  ...
  "rooms": {
    "knock": {
      "!abcdefghijklmo:example.com": {
        "knock_state": {
          "events": []
        }
      }
    }
  },
  ...
```

But that still gives you enough information to know that you have a pending knock for a room.

Reply:
Aaaand afterwards it was brought up that including the membership event is useful for two reasons:

1. It gives clients a way to have a "you knocked on this room" event which they can display in room previews:

    ![image](https://user-images.githubusercontent.com/1342360/100247790-a0592880-2f32-11eb-9589-fe6b553f06bd.png)

    For instance, that "Manolo invited you" text is the preview of the invite membership event that gets inserted into the room. Now this is a bit superficial in that Element iOS could change their code around to support it, and probably wouldn't be enough to convince me to keep the membership in, buuuut...

2. It gives clients a *timestamp* for when the knock occurred, which can actually be quite useful when displaying a list of pending knocks.

I've changed my mind after the above, and think we should keep it in.

Reply:
OK, so should this comment be updated at all?